crates/runtime/src/libcalls.rs

```diff
@@ -493,22 +529,28 @@ macro_rules! ensure {
 /// check is here so we don't segfault from Rust. For other configurations,
 /// these checks are required anyways.
 unsafe fn validate_atomic_addr(
-    instance: &Instance,
+    instance: &mut Instance,
     memory: MemoryIndex,
     addr: u64,
     access_size: u64,
     access_alignment: u64,
-) -> Result<(), TrapCode> {
+) -> Result<Arc<ParkingSpot>, TrapCode> {
     debug_assert!(access_alignment.is_power_of_two());
     ensure!(addr % access_alignment == 0, TrapCode::HeapMisaligned);
 
-    let length = u64::try_from(instance.get_memory(memory).current_length()).unwrap();
+    let mem = instance.get_memory(memory);
+
+    let length = u64::try_from(mem.current_length()).unwrap();
     ensure!(
         addr.saturating_add(access_size) < length,
         TrapCode::HeapOutOfBounds
     );
 
-    Ok(())
+    let defined_index = instance.module().defined_memory_index(memory).unwrap();
+    let mem = &mut *instance.get_defined_memory(defined_index);
+    let spot = mem.as_shared_memory().unwrap().parking_spot();
```

Comment:
I think there's a few issues with this that will need updating. Again though this is where I think that tests should be added to exercise all these cases.

1. Switching to `defined_memory_index` only works if the memory is defined locally, not if the memory is imported. This will want to work with imported memories as well.
2. The wait/notify instructions work with unshared memories in addition to shared memories, so the unwrap for a shared memory needs to be handled. To handle this I'd recommend returning an `Option<T>` and handling `None` in each instruction, for example by sleeping for `wait`, perhaps trapping for an infinite sleep, and doing nothing for `notify`.

Reply:
maybe fixed